Cplus ize printf

fromage17 opened the pull request:
Martin, this PR (cplus-ize-printf) has been updated and now supersedes the other PR (fix-clang-tidy-notifs).  
It contains:
a) it contains consts/constexprs for all magic #s, and provides the formulas that underlie computation of the exponent;
b) it addresses all your feedback from PR fix-clang-tidy-notifs;
c) it contains various bug fixes to floating point representations, especially %g use cases;
d) it eliminates an infinite-recursion risk that had been present;
e) it passes all Test Cases, clang-tidy, and cppcheck -- and I added many floating-point Test Cases;
f) it contains the RespiraWorks pre-commit checks;

With your permission, I'll close the old PR (fix-clang-tidy-notifs).  This current PR supersedes it.

Here's what's next, as I see it:
1) Upon your approval, I'll squash commits and rebase to the HEAD of master (the rebase will be a trivial operation as master hasn't changed in awhile);
2) As you specified, I'll put printf and its README into its own dir next to nanopb;
3) I'll make it compatible with c++17, and make it more c++-style;
4) I'll do some needed cleanup of the code, test cases, and README;

Looking forward to your feedback!

fromage17 committed:
clang-tidy fixes; more to go

fromage17 committed:
clang-tidy fixes; more to go

fromage17 committed:
clang-tidy fixes -- 2nd bunch

fromage17 committed:
.clang-tidy

fromage17 committed:
fixing errs/warns found during compilation of test_suite.cpp

fromage17 committed:
compiler-fixes for test_suite.cpp; also changed /dev/null to /dev/stderr in Makefile to stop hiding compiler errors

fromage17 committed:
clean-up the nan/inf code in printf.c;  supplem tests for nan/inf in test_suite.cpp

fromage17 committed:
include printf.cpp instead of printf.c

fromage17 committed:
printf.cpp: start c++-izing casts

fromage17 committed:
cvt all c-style casts to c++

fromage17 committed:
cvt most c-style #defines to const

fromage17 committed:
FLAGS_FOO() as constexpr functions

fromage17 committed:
FLAGS_FOO as constexpr variables

fromage17 committed:
rm printf.c -- printf.cpp is the new incarnation

fromage17 committed:
copy printf.{cpp,h} back to printf repo after testing in Ventilator repo

fromage17 committed:
test::_putchar,printf_idx,printf_buffer,_out_fct

fromage17 committed:
use same printf.* for internal and external repo

fromage17 committed:
mv <cstdio> are others -> inside _putchar()

fromage17 committed:
printf.cpp printf.h: working

fromage17 committed:
printf.cpp printf.h pre-commit

fromage17 committed:
fixed clang-tidy errs in test/test_suite.cpp

fromage17 committed:
cppcheck fixes

fromage17 committed:
test_suite, clang-tidy, cppcheck : all pass.

fromage17 committed:
fixing 9.99e-to-10.00e bug;  also Makefile improvements

fromage17 committed:
Makefile

fromage17 committed:
frac == 0U out, mv 'rescale' up, oob values tests unified for eEgG

fromage17 committed:
oob values tests mv'd to eEgG and fF cases in big switch

fromage17 committed:
mistaken rounding-test; excess_prec

martukas commented:
I am guessing this PR is dependent on the changes carried out in the previous ones, so I won't touch thus just yet until we merge the other ones.

fromage17 committed:
test_suite.cpp: new TCs for prec>exp10 test; printf.cpp: rm old comm-out code

fromage17 committed:
test_suite: prec vs exp10 TCs

fromage17 committed:
printf.cpp: prec -= exp10

fromage17 committed:
test_suite.cpp: %g cases

fromage17 committed:
printf.cpp: implem proper defn of %g ; streamline test-case printing

fromage17 committed:
test_suite.cpp: more %g cases

fromage17 committed:
test_suite.cpp: refine diagnostic cout msg

fromage17 committed:
test_suite.cpp: use proper c++ casts, numeric_limits<>

fromage17 committed:
exp10-deriving constants

fromage17 committed:
printf.cpp: better constexpr names; test_suite.cpp: organize the float test cases

fromage17 committed:
printf.cpp

fromage17 committed:
test_suite.cpp: added %g-to-e tests

fromage17 committed:
test_suite.cpp: refining %g-to-e test cases

fromage17 committed:
calc_frac,calc_exp10,fill_mant: basic calls, don't replace _etoa's call to _ftoa yet

fromage17 committed:
printf.cpp: calc_frac(),calc_exp10,fill_mantissa() basic calls: TC's working

fromage17 committed:
printf.cpp: calc_frac(),calc_exp10,fill_mantissa():  all calls from _etoa() to these

fromage17 committed:
9.9-to-10.0 rollover bug fixed!

fromage17 committed:
printf.cpp, test_suite.cpp: cleanup after fixing rollover-to-exp10 bug

fromage17 committed:
dbging adjust_sigfigs()

fromage17 committed:
test_suite.cpp & adjust_sigfigs():  ALL TC's pass!

fromage17 committed:
adjust_sigfigs():  misc chgs

fromage17 committed:
printf.cpp: calc_*() pass ptrs not refs; constexprs must not depend on non-const math funcs (e.g., log())

fromage17 committed:
testing, clang-tidy, cppcheck fixes

fromage17 committed:
using CaseSpec alias

fromage17 committed:
using CaseSpec alias, pt 2

fromage17 committed:
Test Cases: using struct alias to specify CaseSpec; also Catch2's CHECK()

fromage17 committed:
Test Cases: refine use of CHECK(); cleanups

fromage17 committed:
Test Cases: cleanups

fromage17 committed:
Makefile: flexible targets - yay

fromage17 committed:
Makefile: targ 'run' runs test_suite; test_suite.cpp: prefer CHECK() to REQUIRE() on many tests

fromage17 committed:
Makefile: 'make checks' will build printf/test_suite and run all tests, incl clang-tidy and cppcheck

fromage17 commented:
Martin,
I've updated this PR's branch again --- the Makefile now builds only what's needed, instead of "everything".  Added targets for running clang-tidy, cppcheck, and the test_suite itself.
I will update this branch again with more improvements.  Next:  C++17 compatibility.  
==Scott

fromage17 committed:
make_for_all_stds, Makefile: cleanups; printf.cpp: finish constexprs
